core/drivers/versal_trng.c

```diff
@@ -0,0 +1,1147 @@
+		0x13, 0xec, 0x5f, 0x97, 0x44, 0x17, 0xc4, 0xa7, 0x7e, 0x3d,
+		0x64, 0x5d, 0x19, 0x73, 0x60, 0x81, 0x4f, 0xdc, 0x22, 0x2a,
+		0x90, 0x88, 0x46, 0xee, 0xb8, 0x14, 0xde, 0x5e, 0x0b, 0xdb,
+		0xe0, 0x32, 0x3a, 0x0a, 0x49, 0x06, 0x24, 0x5c, 0xc2, 0xd3,
+		0xac, 0x62, 0x91, 0x95, 0xe4, 0x79, 0xe7, 0xc8, 0x37, 0x6d,
+		0x8d, 0xd5, 0x4e, 0xa9, 0x6c, 0x56, 0xf4, 0xea, 0x65, 0x7a,
+		0xae, 0x08, 0xba, 0x78, 0x25, 0x2e, 0x1c, 0xa6, 0xb4, 0xc6,
+		0xe8, 0xdd, 0x74, 0x1f, 0x4b, 0xbd, 0x8b, 0x8a, 0x70, 0x3e,
+		0xb5, 0x66, 0x48, 0x03, 0xf6, 0x0e, 0x61, 0x35, 0x57, 0xb9,
+		0x86, 0xc1, 0x1d, 0x9e, 0xe1, 0xf8, 0x98, 0x11, 0x69, 0xd9,
+		0x8e, 0x94, 0x9b, 0x1e, 0x87, 0xe9, 0xce, 0x55, 0x28, 0xdf,
+		0x8c, 0xa1, 0x89, 0x0d, 0xbf, 0xe6, 0x42, 0x68, 0x41, 0x99,
+		0x2d, 0x0f, 0xb0, 0x54, 0xbb, 0x16,
+	};
+	size_t i = 0;
+
+	memcpy(sbx1, sb, sizeof(sb));
+	for (i = 0; i < sizeof(sb); i++) {
+		sbx2[i] = (sb[i] << 1) ^ (((sb[i] >> 7) & 1) * 0x1B);
+		sbx3[i] = sbx2[i] ^ sb[i];
+	}
+}
+
+/*
+ * This function implements the Derivative Function by distilling the entropy
+ * available in its input into a smaller number of bits on the output.
+ * - per NIST SP80090A.
+ *
+ * The Block Cipher algorithm follows sections 10.3.2 and 10.3.3 of the
+ * NIST.SP.800-90Ar1 document
+ */
+static void trng_df_algorithm(struct versal_trng *trng, uint8_t *dfout,
+			      uint32_t flag, const uint8_t *pstr)
+{
+	static bool df_init;
+	const uint8_t df_key[DF_KEY_LEN] = {
+		0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
+		17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31
+	};
+	size_t dfin_len = sizeof(struct trng_dfin) + trng->len;
+	uint8_t *inp_blk = NULL;
+	uint8_t *out_blk = NULL;
+	uintptr_t reminder = 0;
+	size_t xfer_len = 0;
+	uint32_t index = 0;
+	uintptr_t src = 0;
+	uintptr_t dst = 0;
+	size_t offset = 0;
+
+	if (!df_init) {
+		trng_df_init();
+		df_init = true;
+	}
+
+	if (flag == DF_SEED)
+		trng->dfin.val2 = TEE_U32_TO_BIG_ENDIAN(TRNG_PERS_STR_LEN);
+	else
+		trng->dfin.val2 = TEE_U32_TO_BIG_ENDIAN(TRNG_GEN_LEN);
+
+	trng->dfin.pad_data[0] = DF_PAD_VAL;
+
+	if (!pstr) {
+		if (trng->len > (MAX_PRE_DF_LEN + TRNG_PERS_STR_LEN))
+			panic();
+
+		dfin_len = dfin_len - TRNG_PERS_STR_LEN - MAX_PRE_DF_LEN;
+		trng->dfin.val1 = TEE_U32_TO_BIG_ENDIAN(trng->len);
+
+		xfer_len = DF_PAD_DATA_LEN;
+		src = (uintptr_t)trng->dfin.pad_data;
+		offset = MAX_PRE_DF_LEN + TRNG_PERS_STR_LEN - trng->len;
+	} else {
+		if (trng->len > MAX_PRE_DF_LEN)
+			panic();
+
+		memcpy(trng->dfin.pstr, pstr, TRNG_PERS_STR_LEN);
+		dfin_len = dfin_len - MAX_PRE_DF_LEN;
+		trng->dfin.val1 = TEE_U32_TO_BIG_ENDIAN(trng->len +
+							TRNG_PERS_STR_LEN);
+		xfer_len = DF_PAD_DATA_LEN + TRNG_PERS_STR_LEN;
+		src = (uintptr_t)trng->dfin.pstr;
+		offset = MAX_PRE_DF_LEN - trng->len;
+	}
+
+	/* Move back into the dfin structure */
+	dst = src - offset;
+	reminder = (uintptr_t)&trng->dfin + sizeof(trng->dfin) - offset;
+	if (offset) {
+		memcpy((void *)dst, (void *)src, xfer_len);
```

Comment:
`memove()` seems a better option.

Reply:
um, why, how do you know src is not being used somewhere else. I dont feel comfortable with this change. why do you feel it is required?

Reply:
src and dst are in the same address range. As stated in the inline comment, this move data inside a single buffer, hence this should use `memmove()` IIUC.

Reply:
yes, i was thinking of overlaps. From memcpy() man page:
> DESCRIPTION
>      The  memcpy()  function  copies n bytes from memory area src to memory area dest.
>      **The memory areas must not overlap.  Use memmove(3) if the memory areas do overlap.**

Maybe `Xil_SMemCpy()` in the referred code ensures destination address is lower than source when buffer do overlap. Using `memcpy()` you do not have this ensurance.

Reply:
um, I think you are right.
maybe we should just do likewise?
```
int Xil_SMemCpy(void *Dest, const u32 DestSize,
	const void *Src, const u32 SrcSize, const u32 CopyLen)
{
	int Status = XST_FAILURE;
	const u8 *Src8 = (const u8 *) Src;
	const u8 *Dst8 = (u8 *) Dest;
	void * volatile DestTemp = Dest;
	const void * volatile SrcTemp = Src;

	if ((Dest == NULL) || (Src == NULL)) {
		Status =  XST_INVALID_PARAM;
	}
	else if ((CopyLen == 0U) || (DestSize < CopyLen) || (SrcSize < CopyLen)) {
		Status =  XST_INVALID_PARAM;
	}
	/* Return error for overlap string */
	else if ((Src8 < Dst8) && (&Src8[CopyLen - 1U] >= Dst8)) {
		Status =  XST_INVALID_PARAM;
	}
	else if ((Dst8 < Src8) && (&Dst8[CopyLen - 1U] >= Src8)) {
		Status =  XST_INVALID_PARAM;
	}
	else {
		(void)memcpy(DestTemp, SrcTemp, CopyLen);
		Status = XST_SUCCESS;
	}

	return Status;
}
```

Reply:
I think when I rewrote this code early this year I already made all the proper corrections thought so we are probably are going in circles...but it is a good thing to review.

Reply:
just pushed a safeguard (in any case the algorithm expected the remainder to be set to 0 so truncating seems to be fine.